pkg/providers/vsphere/vmprovider_vm_utils.go

```diff
@@ -46,24 +49,81 @@ func errToConditionReasonAndMessage(err error) (string, string) {
 
 func GetVirtualMachineClass(
 	vmCtx pkgctx.VirtualMachineContext,
-	k8sClient ctrlclient.Client) (*vmopv1.VirtualMachineClass, error) {
+	k8sClient ctrlclient.Client) (vmopv1.VirtualMachineClass, error) {
 
-	key := ctrlclient.ObjectKey{Name: vmCtx.VM.Spec.ClassName, Namespace: vmCtx.VM.Namespace}
-	vmClass := &vmopv1.VirtualMachineClass{}
-	if err := k8sClient.Get(vmCtx, key, vmClass); err != nil {
+	if vmopv1util.IsClasslessVM(*vmCtx.VM) &&
+		pkgcfg.FromContext(vmCtx).Features.VMImportNewNet {
+
+		// If VM was not deployed from a class, then synthesize the class from
+		// the underlying VM.
+		vmCtx.Logger.Info("Synthesize class from underlying VM")
+		return getVirtualMachineClassFromVM(vmCtx)
+	}
+
+	obj, err := getVirtualMachineClassFromClassName(vmCtx, k8sClient)
+
+	if err != nil {
 		reason, msg := errToConditionReasonAndMessage(err)
-		conditions.MarkFalse(vmCtx.VM, vmopv1.VirtualMachineConditionClassReady, reason, msg)
-		return nil, err
+		conditions.MarkFalse(
+			vmCtx.VM, vmopv1.VirtualMachineConditionClassReady, reason, msg)
+		return vmopv1.VirtualMachineClass{}, err
 	}
 
 	conditions.MarkTrue(vmCtx.VM, vmopv1.VirtualMachineConditionClassReady)
 
-	return vmClass, nil
+	return obj, nil
+}
+
+func getVirtualMachineClassFromVM(
+	vmCtx pkgctx.VirtualMachineContext) (vmopv1.VirtualMachineClass, error) {
+
+	obj := vmopv1.VirtualMachineClass{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: vmCtx.VM.Spec.ClassName,
```

Comment:
IIUC, this function will be called when the VM spec doesn't have a class name (i.e. `IsClasslessVM()` returns true). So I am not sure what the point of setting an empty name here is. Did you mean to set the namespace instead? If not, should we consider adding the namespace here as well?